rtz/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+    ///
+    /// Returns None if another RTZ instance already exists in this thread.
+    ///
+    /// # Safety
+    ///
+    /// This function uses inline assembly to modify the FPCR register. The
+    /// operations are safe when used as intended through this API, as it
+    /// maintains the following invariants:
+    /// - Only one instance can exist per thread
+    /// - The previous FPCR value is always restored on drop
+    /// - The type cannot be sent between threads
+    #[cfg(target_arch = "aarch64")]
+    #[inline]
+    pub fn set() -> Option<RTZ> {
+        // Try to acquire ownership of FPCR
+        let state = FPCR_OWNED.with(|owned| {
+            let observed_state = owned.get();
+            if observed_state == FPCRState::Idle {
+                owned.set(FPCRState::Active);
+            }
+            observed_state
+        });
+
+        match state {
+            FPCRState::Idle => {
+                let mut prev_fpcr: u64;
+                unsafe {
+                    // Read current FPCR value and set round-toward-zero mode
+                    core::arch::asm!(
+                        // Read current FPCR value
+                        "mrs {prev_fpcr}, fpcr",
+                        "orr {tmp}, {prev_fpcr}, {rmode}",
+                        "msr fpcr, {tmp}",
+                        prev_fpcr = out(reg) prev_fpcr,
+                        tmp = out(reg) _,
+                        rmode = const FPCR_RMODE_BITS,
+                    );
+                }
+
+                Some(Self {
+                    prev_fpcr,
+                    _no_send: PhantomData,
+                })
+            }
+            FPCRState::Active => None,
+        }
+    }
+
+    #[cfg(not(target_arch = "aarch64"))]
+    #[inline]
+    fn set() -> Option<RTZ> {
+        unimplemented!()
+    }
+
+    /// Reads the current value of the FPCR register.
+    ///
+    /// This method is primarily intended for debugging and verification
+    /// purposes.
+    #[cfg(target_arch = "aarch64")]
+    #[inline]
+    pub fn read(&self) -> u64 {
+        let mut value: u64;
+        unsafe {
+            core::arch::asm!(
+                "mrs {}, fpcr",
+                out(reg) value,
+                options(nomem, nostack, preserves_flags)
+            );
+        }
+        value
+    }
+
+    #[cfg(not(target_arch = "aarch64"))]
+    #[inline]
+    pub fn read(&self) -> u64 {
+        unimplemented!()
+    }
+
+    /// Writes a new value to the FPCR register.
+    ///
+    /// This is a low-level operation that directly modifies the FPCR register.
+    /// It should be used with caution as improper values can affect
+    /// floating-point behavior.
+    ///
+    /// # Safety
+    ///
+    /// This operation is safe because:
+    /// - The RTZ instance proves we have exclusive access to FPCR
+    /// - The write operation is atomic
+    #[cfg(target_arch = "aarch64")]
+    #[inline]
+    pub fn write(&self, value: u64) {
```

Comment:
How do we know for sure the compiler won't re-order instructions around this? Is that guaranteed by and `asm!` block? If not, are the barriers we can add here to force the compiler?

Reply:
The `asm!` block provides no guarantee it's the lifetime that does it. It ensures that the drop of RTZ, which calls write, will be after the last use of RTZ. It's modelled after how lifetimes of pointers around FFI are tracked. There you don't want to have your free suddenly be in front of the call using the pointer.  

I'll look for official documentation or code in std to substantiate this design.

Reply:
I'm exploring other options for this. It turns out that the lifetime by itself is not sufficient under certain inlining conditions. In a version without Drop I've been able to let the compiler place the restore of FPCR before the dependant computation.